src/group_theory/group_ring.lean

```diff
@@ -0,0 +1,216 @@
+  (f : P ≃ₗ[k] P') (H : ∀ g x, f (of k G g • x) = of k G g • f x) :
+  P ≃ₗ[group_algebra k G] P' :=
+{ ..f.to_equiv, ..mk_linear f.to_linear_map H }
+
+end
+
+variables {k : Type*} [comm_semiring k] {G : Type*} [group G] {n : ℕ}
+
+instance {H : Type*} [mul_action G H] : distrib_mul_action G (group_algebra k H) :=
+finsupp.comap_distrib_mul_action
+
+/-- If `k[H]` is a `k[G]`-module, then a `k`-linear map `f: k[H] →ₗ P` of `k[G]`-modules which
+satisfies `f(g • h) = g • f(h)` for all `g ∈ G, h ∈ H` is `k[G]`-linear. -/
+lemma map_smul_of_map_smul_of {H : Type*} [monoid H]
+  [module (group_algebra k G) (group_algebra k H)] {P : Type*} [add_comm_monoid P] [module k P]
+  [module (group_algebra k G) P] [is_scalar_tower k (group_algebra k G) (group_algebra k H)]
+  [is_scalar_tower k (group_algebra k G) P] (f : group_algebra k H →ₗ[k] P)
+  (h : ∀ (g : G) (x : H), f (of k G g • of k H x) = of k G g • f (of k H x))
+  (g : group_algebra k G) (x : group_algebra k H) : f (g • x) = g • f x :=
+begin
+  rw mk_linear_smul,
+  intros a b,
+  refine b.induction_on (by exact h a) _ _,
+  { intros s t hs ht,
+    simp only [smul_add, f.map_add, hs, ht] },
+  { intros r s hs,
+    simp only [smul_algebra_smul_comm, f.map_smul, hs] }
+end
+
+instance : smul_comm_class k G (group_algebra k (fin n → G)) :=
+{ smul_comm := λ x y z,
+  begin
+    refine z.induction_on _ _ _,
+    { intro z,
+      simp only [finsupp.smul_single', finsupp.comap_smul_single, of_apply] },
+    { intros f g hf hg,
+      simp only [hf, hg, smul_add] },
+    { intros r f hf,
+      ext,
+      simp only [finsupp.comap_smul_apply, finsupp.coe_smul, pi.smul_apply]}
+  end }
+
+instance : module (group_algebra k G) (group_algebra k (fin n → G)) :=
+action_to_module
+
+lemma single_smul_single (g : G) (x : fin n → G) (r s : k) :
+  ((•) : group_algebra k G → group_algebra k (fin n → G) → group_algebra k (fin n → G))
+  (finsupp.single g r) (finsupp.single x s) =
+  finsupp.single (g • x) (r * s) :=
+begin
+  show finsupp.total _ _ _ _ _ = _,
+  simp only [finsupp.smul_single', finsupp.total_single, finsupp.comap_smul_single],
+end
+
+lemma of_smul_of (g : G) (x : fin n → G) :
+  of k G g • of k (fin n → G) x = of k (fin n → G) (g • x) :=
+show finsupp.total _ _ _ _ _ = _, by simp
+
+instance is_scalar_tower' : is_scalar_tower k (group_algebra k G) (group_algebra k (fin n → G)) :=
+{ smul_assoc := λ x y z,
+  begin
+    refine y.induction_on _ _ _,
+    { intro y,
+      refine z.induction_on _ _ _,
+      { intro z,
+        simp only [single_smul_single, mul_one, finsupp.smul_single', of_apply] },
+      { intros f g hf hg,
+        simp only [hf, hg, smul_add] },
+      { intros r f hf,
+        simp only [smul_def, mul_one, finsupp.smul_single', one_smul,
+          finsupp.total_single, of_apply] at hf ⊢}},
+    { intros f g hf hg,
+      simp only [smul_add, add_smul, hf, hg] },
+    { intros r f hf,
+      simp only [smul_def, ring_hom.id_apply, linear_map.map_smulₛₗ] }
```

Comment:
For what it's worth, this golfs to:
```suggestion
    refine y.induction_on
      (λ w, _)
      (λ f g hf hg, by simp only [smul_add, add_smul, hf, hg])
      (λ r f hf, by simp only [smul_def, ring_hom.id_apply, linear_map.map_smulₛₗ]),
    exact z.induction_on
      (λ v, by simp only [mul_one, of_apply, finsupp.smul_single', single_smul_single])
      (λ f g hf hg, by simp only [hf, hg, smul_add])
      (λ r f hf, by simp only [smul_def, mul_one, one_smul, of_apply, finsupp.total_single,
        finsupp.smul_single']),
```
but I think it is more important to address Eric's suggestions above.